lib/perl/Genome/InstrumentData/AlignmentResult/Merged.pm

```diff
@@ -707,4 +786,50 @@ sub scalar_property_from_underlying_alignment_results {
 }
 
 
+sub _lock_per_lane_alignment {
+    my $self = shift;
+
+    for my $alignment ($self->collect_individual_alignments) {
+        my $id = $alignment->id;
+        unless ($alignment->get_merged_alignment_results) {
+            my @bams = $alignment->alignment_bam_file_paths;
+            unless (@bams) {
+               die $self->error_message("Per lane alignment $id has neither merged results nor bam paths. Please create an apipe-support ticket for this.");
+               #There is no way to recreate per lane bam if merged bam
+               #does not exist and per lane bam is removed. Software
+               #result of this per lane alignment needs to be removed and
+               #this per lane instrument data needs to be realigned
+               #with that aligner.
+            }
+
+            unless ($ENV{UR_DBI_NO_COMMIT}) {
+                my $lock_var = 'genome_instrument_data_alignment_result-merged-per-lane-'.$id.'/lock';
+                my $lock = Genome::Sys->lock_resource(
```

Comment:
As discussed in the code review today, this could cause a deadlock when some number of builds are in contention for 2+ shared instrument data. If every lock had information about other locks, they could release/retry in a random amount of time if a deadlock is detected. Or there could be a tiebreaker where the lower numbered instrument data in a sort is allowed to have the lock first, and other builds release and retry in an hour. Something similar.

This potentially should be part of the locking system?